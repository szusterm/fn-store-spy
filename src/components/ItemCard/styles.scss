@import '../../scss/variables';
@import '../../scss/shadow';
@import '../../scss/transition';

.item-card {
	$strip-height: 16px;

	cursor: pointer;

	max-width: 100%;
	height: auto;
	margin: 10px;

	&--top-part, &--bot-part {
		background-color: $body-bg;
		@include shadow(sm);

		@include transition(sh);
		transition-property: box-shadow;
	}

	&:hover {
		.item-card--top-part, .item-card--bot-part {
			@include shadow(md);
		}
	}

	&--top-part {
		padding: 12px;

		border-top-left-radius: 50%;
		border-top-right-radius: 50%;

		position: relative;

		&::after {
			width: 100%;
			background: $body-bg;
			height: $strip-height;
			content: "";

			display: block;
			position: absolute;
			top: calc(100% - #{$strip-height}/2);
			left: 0;
			right: 0;
		}
	}

	&--bot-part {
		$padding: 15px;

		height: 110px;
		padding-top: $strip-height/2;

		border-bottom-left-radius: 10px;
		border-bottom-right-radius: 10px;

		display: flex;
		flex-direction: column;
		justify-content: space-between;
		align-items: center;

		.bot-part {
			color: $body-color;
			text-align: center;
			font-family: 'Roboto', Arial;

			&--name {
				font-size: 1.1em;
				font-weight: 500;
				text-align: center;
				padding: 0 $padding;
			}

			&--description {
				width: 100%;
				padding: $padding;

				display: flex;
				flex-direction: row;
				justify-content: space-around;
				align-items: center;

				.description {
					&--type {
						font-size: 0.8em;
						font-weight: 400;
					}

					&--price {
						font-size: 1em;
						font-weight: 100;
					}
				}
			}
		}
	}
}