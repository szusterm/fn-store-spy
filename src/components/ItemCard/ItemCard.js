import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {addItemToOrder, setActiveInOrderedItem} from '../../redux/actions';

import './styles.scss';

export class ItemCard extends Component {
	constructor() {
		super();

		this.switchOrder = this.switchOrder.bind(this);
	}

	switchOrder() {
		const {_id, name, rarity} = this.props;

		if (this.checkThatItemIsOrdered()) {
			const item = {
				_id,
				name,
				rarity
			};

			this.props.addItemToOrder(item);
		}
		else {
			const indexInOrder = this.getIndexInOrder();
			//method to remove item from order
		}
	}

	checkThatItemIsOrdered() {
		const {_id} = this.props;
		const {orderedItems} = this.props;

		const sameItemInOrder = orderedItems.find((orderedItem) => (_id === orderedItem._id));

		return (sameItemInOrder !== undefined);
	}

	getIndexInOrder() {
		const {_id} = this.props;
		const {orderedItems} = this.props;

		let indexInOrder = null;

		orderedItems.find((orderedItem, orderedItemIndex) => {
			if (_id === orderedItem._id) {
				indexInOrder = orderedItemIndex;
			}
		});

		return indexInOrder;
	}

	render() {
		const {name, price, rarity, type, imageSrc} = this.props;

		return (
			<div
				className={'item-card item-card--raised'}
				onClick={this.switchOrder}
			>
				<div className={'item-card--top-part'}>
					<div className={`top-part--img-box top-part--img-box--${rarity}`}>
						<img
							className={'img-box--image'}
							src={imageSrc}
						/>
					</div>
				</div>
				<div className={'item-card--bot-part'}>
					<div className={'bot-part--name'}>{name}</div>
					{
						(this.checkThatItemIsOrdered()) &&
							<span>ORDERED</span>
					}
					<div className={'bot-part--description'}>
						<div className={'description--type'}>{type}</div>
						<div className={'description--price'}>{price}</div>
					</div>
				</div>
			</div>
		);
	}
}

ItemCard.propTypes = {
	_id: PropTypes.string.isRequired,
	name: PropTypes.string.isRequired,
	price: PropTypes.string.isRequired,
	rarity: PropTypes.string.isRequired,
	type: PropTypes.string.isRequired,
	imageSrc: PropTypes.string.isRequired,
	orderedItems: PropTypes.array, //redux,
	addItemToOrder: PropTypes.func, //redux
	setActiveInOrderedItem: PropTypes.func //redux
};

const mapStateToProps = (state) => {
	const {items: orderedItems} = state.order;
	return {orderedItems};
};

const mapDispatchToProps = {addItemToOrder, setActiveInOrderedItem};

export default connect(mapStateToProps, mapDispatchToProps)(ItemCard);