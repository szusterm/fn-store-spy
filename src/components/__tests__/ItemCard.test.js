import React from 'react';
import {shallow} from 'enzyme';

import ItemCard from '../ItemCard';

const setup = (propOverrides) => {
	const props = Object.assign({
		name: 'Super Axe',
		price: 2000,
		rarity: 'uncommon',
		imgSrc: 'https://example.com'
	}, propOverrides);

	const wrapper = shallow(<ItemCard {...props}/>);

	return {props, wrapper};
};

describe('ItemCard Component', () => {
	it('renders correctly', () => {
		const {wrapper} = setup();

		expect(wrapper).toMatchSnapshot();
	});
});