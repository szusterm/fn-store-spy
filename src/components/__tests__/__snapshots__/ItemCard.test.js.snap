// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemSearchCard Component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ItemSearchCard
    imageSrc="https://example.com"
    name="Super Axe"
    price="2000"
    rarity="uncommon"
    type="axe"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="item-card--top-part"
        >
          <div
            className="top-part--img-box top-part--img-box--uncommon"
          >
            <img
              className="img-box--image"
              src="https://example.com"
            />
          </div>
        </div>,
        <div
          className="item-card--bot-part"
        >
          <div
            className="bot-part--name"
          >
            Super Axe
          </div>
          <div
            className="bot-part--description"
          >
            <div
              className="description--type"
            >
              axe
            </div>
            <div
              className="description--price"
            >
              2000
            </div>
          </div>
        </div>,
      ],
      "className": "item-card item-card--raised",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="top-part--img-box top-part--img-box--uncommon"
          >
            <img
              className="img-box--image"
              src="https://example.com"
            />
          </div>,
          "className": "item-card--top-part",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              className="img-box--image"
              src="https://example.com"
            />,
            "className": "top-part--img-box top-part--img-box--uncommon",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "img-box--image",
              "src": "https://example.com",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="bot-part--name"
            >
              Super Axe
            </div>,
            <div
              className="bot-part--description"
            >
              <div
                className="description--type"
              >
                axe
              </div>
              <div
                className="description--price"
              >
                2000
              </div>
            </div>,
          ],
          "className": "item-card--bot-part",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Super Axe",
              "className": "bot-part--name",
            },
            "ref": null,
            "rendered": "Super Axe",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="description--type"
                >
                  axe
                </div>,
                <div
                  className="description--price"
                >
                  2000
                </div>,
              ],
              "className": "bot-part--description",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "axe",
                  "className": "description--type",
                },
                "ref": null,
                "rendered": "axe",
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "2000",
                  "className": "description--price",
                },
                "ref": null,
                "rendered": "2000",
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="item-card--top-part"
          >
            <div
              className="top-part--img-box top-part--img-box--uncommon"
            >
              <img
                className="img-box--image"
                src="https://example.com"
              />
            </div>
          </div>,
          <div
            className="item-card--bot-part"
          >
            <div
              className="bot-part--name"
            >
              Super Axe
            </div>
            <div
              className="bot-part--description"
            >
              <div
                className="description--type"
              >
                axe
              </div>
              <div
                className="description--price"
              >
                2000
              </div>
            </div>
          </div>,
        ],
        "className": "item-card item-card--raised",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="top-part--img-box top-part--img-box--uncommon"
            >
              <img
                className="img-box--image"
                src="https://example.com"
              />
            </div>,
            "className": "item-card--top-part",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                className="img-box--image"
                src="https://example.com"
              />,
              "className": "top-part--img-box top-part--img-box--uncommon",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "img-box--image",
                "src": "https://example.com",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="bot-part--name"
              >
                Super Axe
              </div>,
              <div
                className="bot-part--description"
              >
                <div
                  className="description--type"
                >
                  axe
                </div>
                <div
                  className="description--price"
                >
                  2000
                </div>
              </div>,
            ],
            "className": "item-card--bot-part",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Super Axe",
                "className": "bot-part--name",
              },
              "ref": null,
              "rendered": "Super Axe",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div
                    className="description--type"
                  >
                    axe
                  </div>,
                  <div
                    className="description--price"
                  >
                    2000
                  </div>,
                ],
                "className": "bot-part--description",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "axe",
                    "className": "description--type",
                  },
                  "ref": null,
                  "rendered": "axe",
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "2000",
                    "className": "description--price",
                  },
                  "ref": null,
                  "rendered": "2000",
                  "type": "div",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
